opm/simulators/wells/RateConverter.hpp

```diff
@@ -488,14 +516,23 @@ namespace Opm {
                     auto& saltConcentration = ra.saltConcentration;
 
                     // sum p, rs, rv, and T.
-                    double hydrocarbonPV = pv_cell*hydrocarbon;
-                    if (hydrocarbonPV > 0) {
-                        pv += hydrocarbonPV;
-                        p += fs.pressure(FluidSystem::oilPhaseIdx).value()*hydrocarbonPV;
-                        rs += fs.Rs().value()*hydrocarbonPV;
-                        rv += fs.Rv().value()*hydrocarbonPV;
-                        T += fs.temperature(FluidSystem::oilPhaseIdx).value()*hydrocarbonPV;
-                        saltConcentration += fs.saltConcentration().value()*hydrocarbonPV;
+                    if (has_hc) {
+                        double hydrocarbonPV = pv_cell*hydrocarbon;
+                        if (hydrocarbonPV > 0) {
```

Comment:
Just wondering, whether it would be possible to skip the additional grid traversal?

- Do not change pv, p, etc directly in the loop
- Instead use two temporary versions of the quantities (one for has_hc being true, the other for false) and compute both versions.
- In the same loop also compute has_hc
- After the loop assign the correct versions of the temporaries based on has_hc to the real variable pv, p, etc.

Reply:
I agree, that would be better. Note though that the loop will usually terminate very quickly, except in the cases where the code currently fails.

I will adapt to your suggestion.

Reply:
Just a note: I need two temporary versions of the quantities for each region. So it is a little more involved.

Reply:
I will base the new version on #1961 which already is almost there.